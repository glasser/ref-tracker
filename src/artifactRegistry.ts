import * as core from '@actions/core';
import { ArtifactRegistryClient } from '@google-cloud/artifact-registry';

export interface GetAllEquivalentTagsOptions {
  /** The name of the specific Docker image in question (ie, a Docker
   * "repository", not an Artifact Registry "repository" that contains them.) */
  dockerImageRepository: string;
  tag: string;
}

export interface DockerRegistryClient {
  getAllEquivalentTags(options: GetAllEquivalentTagsOptions): Promise<string[]>;
}

export class ArtifactRegistryDockerRegistryClient {
  private client: ArtifactRegistryClient;
  constructor(
    /** A string of the form
     * `projects/PROJECT/locations/LOCATION/repositories/REPOSITORY`; this is an
     * Artifact Registry repository, which is a set of Docker repositories. */
    private artifactRegistryRepository: string,
  ) {
    this.client = new ArtifactRegistryClient();
  }

  async getAllEquivalentTags({
    dockerImageRepository,
    tag,
  }: GetAllEquivalentTagsOptions): Promise<string[]> {
    // Input is relatively trusted; this is largely to prevent mistaken uses
    // like specifying a full Docker-style repository with slashes.
    if (dockerImageRepository.includes('/')) {
      throw Error('repository cannot contain a slash');
    }
    if (tag.includes('/')) {
      throw Error('tag cannot contain a slash');
    }

    const tagPrefix = `${this.artifactRegistryRepository}/packages/${dockerImageRepository}/tags/`;

    core.info(`Fetching tag ${tagPrefix}${tag}`);
    // Note: this throws if the repository or tag are not found.
    const { version } = (
      await this.client.getTag({ name: `${tagPrefix}${tag}` })
    )[0];
    if (!version) {
      throw Error(`No version found for ${dockerImageRepository}@${tag}`);
    }

    core.info(`Fetching version ${version}`);
    const { relatedTags } = (
      await this.client.getVersion({ name: version, view: 'FULL' })
    )[0];
    if (!relatedTags) {
      throw Error(`No related tags returned for ${version}`);
    }
    return relatedTags.map(({ name }) => {
      if (!name?.startsWith(tagPrefix)) {
        throw Error(
          `Expected tag name to start with '${tagPrefix}'; got '${name}'`,
        );
      }
      return name.slice(tagPrefix.length);
    });
  }
}
