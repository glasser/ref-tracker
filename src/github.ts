import { getOctokit } from '@actions/github';
import * as core from '@actions/core';

export interface ResolveRefToShaOptions {
  repoURL: string;
  ref: string;
}

export interface GetTreeSHAForPathOptions {
  repoURL: string;
  ref: string;
  path: string;
}

export interface GitHubClient {
  resolveRefToSha(options: ResolveRefToShaOptions): Promise<string>;
  getTreeSHAForPath(options: GetTreeSHAForPathOptions): Promise<string>;
}

interface OwnerAndRepo {
  owner: string;
  repo: string;
}

function parseRepoURL(repoURL: string): OwnerAndRepo {
  const m = repoURL.match(/\bgithub\.com\/([\w.-]+)\/([\w.-]+?)(?:\.git|\/)?$/);
  if (!m) {
    throw Error(`Can only track GitHub repoURLs, not ${repoURL}`);
  }
  return { owner: m[1], repo: m[2] };
}

export class OctokitGitHubClient {
  constructor(private octokit: ReturnType<typeof getOctokit>) {}

  async resolveRefToSha({
    repoURL,
    ref,
  }: ResolveRefToShaOptions): Promise<string> {
    const { owner, repo } = parseRepoURL(repoURL);
    const sha = (
      await this.octokit.rest.repos.getCommit({
        owner,
        repo,
        ref,
        mediaType: {
          format: 'sha',
        },
      })
    ).data as unknown;
    // The TS types don't understand that `mediaType: {format: 'sha'}` turns
    // `.data` into a string, so we have to cast to `unknown` and check
    // ourselves.
    if (typeof sha !== 'string') {
      throw Error('Expected string response');
    }
    return sha;
  }

  async getTreeSHAForPath({
    repoURL,
    ref,
    path,
  }: GetTreeSHAForPathOptions): Promise<string> {
    const { owner, repo } = parseRepoURL(repoURL);
    // FIXME error handling
    const content = await this.octokit.rest.repos.getContent({
      owner,
      repo,
      ref,
      path,
      mediaType: {
        format: 'object',
      },
    });
    core.info(`Got content ${JSON.stringify(content, null, 2)}`);
    return 'yay';
  }
}
