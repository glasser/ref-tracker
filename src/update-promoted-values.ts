import * as core from '@actions/core';
import { RE2 } from 're2-wasm';
import * as yaml from 'yaml';
import { ScalarTokenWriter, getTopLevelBlocks } from './yaml';

// FIXME we will want to make a way to filter promote-ables so that we can
// separate "promote to staging" from "promote to dev" for pipelines that have
// multiple

interface Promote {
  scalarTokenWriter: ScalarTokenWriter;
  value: string;
}

export async function updatePromotedValues(
  contents: string,
  promotionTargetRegexp: string | null,
): Promise<string> {
  return core.group('Processing promote', async () => {
    // We use re2-wasm instead of built-in RegExp so we don't have to worry about
    // REDOS attacks.
    const promotionTargetRE2 = promotionTargetRegexp
      ? new RE2(promotionTargetRegexp, 'u')
      : null;

    // See updateGitRefs for an explanation about our approach to parsing YAML.
    core.info('Parsing');
    const topLevelTokens = [...new yaml.Parser().parse(contents)];
    const documents = [
      ...new yaml.Composer({ keepSourceTokens: true }).compose(topLevelTokens),
    ];

    // We decide what to do and then we do it, just in case there are any
    // overlaps between our reads and writes.
    core.info('Looking for promote');
    const promotes = documents.flatMap((document) =>
      findPromotes(document, promotionTargetRE2),
    );

    core.info('Copying values');
    for (const { scalarTokenWriter, value } of promotes) {
      scalarTokenWriter.write(value);
    }
    core.info('Stringifying');
    return topLevelTokens
      .map((topLevelToken) => yaml.CST.stringify(topLevelToken))
      .join('');
  });
}

function findPromotes(
  document: yaml.Document.Parsed,
  promotionTargetRE2: RE2 | null,
): Promote[] {
  const { blocks } = getTopLevelBlocks(document);
  const promotes: Promote[] = [];
  for (const [myName, me] of blocks) {
    if (promotionTargetRE2 && !promotionTargetRE2.test(myName)) {
      continue;
    }
    if (!me.has('promote')) {
      continue;
    }
    const promote = me.get('promote');
    if (!yaml.isMap(promote)) {
      throw Error(`The value at ${myName}.promote must be a map`);
    }
    const from = promote.get('from');
    if (typeof from !== 'string') {
      throw Error(`The value at ${myName}.promote.from must be a string`);
    }
    const fromBlock = blocks.get(from);
    if (!fromBlock) {
      throw Error(
        `The value at ${myName}.promote.from must reference a top-level key with map value`,
      );
    }

    // FIXME use default yamlPaths
    const yamlPathsSeq = promote.get('yamlPaths');
    if (!yaml.isSeq(yamlPathsSeq)) {
      throw Error(`The value at ${myName}.promote.yamlPaths must be an array`);
    }
    const yamlPaths = yamlPathsSeq.toJSON();
    if (!Array.isArray(yamlPaths)) {
      throw Error('YAMLSeq.toJSON surprisingly did not return an array');
    }
    if (!yamlPaths.every(isCollectionPath)) {
      throw Error(
        `The value at ${myName}.promote.yamlPaths must be an array whose elements are arrays of strings or numbers`,
      );
    }

    for (const collectionPath of yamlPaths) {
      const sourceValue = fromBlock.getIn(collectionPath);
      if (typeof sourceValue !== 'string') {
        throw Error(`Could not promote from ${[from, ...collectionPath]}`);
      }
      // true means keepScalar, ie get the scalar node to write.
      const targetNode = me.getIn(collectionPath, true);
      if (!yaml.isScalar(targetNode)) {
        throw Error(`Could not promote to ${[myName, ...collectionPath]}`);
      }
      const scalarToken = targetNode.srcToken;
      if (!yaml.CST.isScalar(scalarToken)) {
        // this probably can't happen, but let's make the types happy
        throw Error(
          `${[myName, ...collectionPath]} value must come from a scalar token`,
        );
      }
      promotes.push({
        scalarTokenWriter: new ScalarTokenWriter(scalarToken, document.schema),
        value: sourceValue,
      });
    }
  }
  return promotes;
}

type CollectionPath = CollectionIndex[];
type CollectionIndex = string | number;

function isCollectionPath(value: unknown): value is CollectionPath {
  return Array.isArray(value) && value.every(isCollectionIndex);
}

function isCollectionIndex(value: unknown): value is CollectionIndex {
  return typeof value === 'string' || typeof value === 'number';
}
