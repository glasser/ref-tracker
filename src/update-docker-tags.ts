import * as core from '@actions/core';
import min from 'lodash/min';
import * as yaml from 'yaml';
import { DockerRegistryClient } from './artifactRegistry';
import {
  ScalarTokenWriter,
  getStringAndScalarTokenFromMap,
  getStringValue,
  getTopLevelBlocks,
} from './yaml';

interface Trackable {
  trackMutableTag: string;
  dockerImageRepository: string;
  tag: string;
  tagScalarTokenWriter: ScalarTokenWriter;
}

export async function updateDockerTags(
  contents: string,
  dockerRegistryClient: DockerRegistryClient,
): Promise<string> {
  return core.group('Processing trackMutableTag', async () => {
    // See updateGitRefs for an explanation about our approach to parsing YAML.
    core.info('Parsing');
    const topLevelTokens = [...new yaml.Parser().parse(contents)];
    const documents = [
      ...new yaml.Composer({ keepSourceTokens: true }).compose(topLevelTokens),
    ];

    if (documents.length > 1) {
      throw new Error('Multiple documents in YAML file');
    }

    if (documents.length < 1) {
      return contents;
    }

    const document = documents[0];

    core.info('Looking for trackMutableTag');
    const trackables = findTrackables(document);

    core.info('Checking tags against Artifact Registry');
    await checkTagsAgainstArtifactRegistryAndModifyScalars(
      trackables,
      dockerRegistryClient,
    );
    core.info('Stringifying');
    return topLevelTokens
      .map((topLevelToken) => yaml.CST.stringify(topLevelToken))
      .join('');
  });
}

function findTrackables(doc: yaml.Document.Parsed): Trackable[] {
  const trackables: Trackable[] = [];

  const { blocks, globalBlock } = getTopLevelBlocks(doc);

  let globalDockerImageRepository: string | null = null;

  if (globalBlock?.has('dockerImage')) {
    const dockerImageBlock = globalBlock.get('dockerImage');
    if (!yaml.isMap(dockerImageBlock)) {
      throw Error('Document has `global.dockerImageBlock` that is not a map');
    }
    // Read repository from 'global' (keeping it null if it's not
    // there, though throwing if it's there as non-strings).
    globalDockerImageRepository = getStringValue(
      dockerImageBlock,
      'repository',
    );
  }

  for (const [key, value] of blocks) {
    if (!value.has('dockerImage')) {
      continue;
    }
    const dockerImageBlock = value.get('dockerImage');
    if (!yaml.isMap(dockerImageBlock)) {
      throw Error(`Document has \`${key}.dockerImage\` that is not a map`);
    }

    const dockerImageRepository =
      getStringValue(dockerImageBlock, 'repository') ??
      globalDockerImageRepository;
    const trackMutableTag = getStringValue(dockerImageBlock, 'trackMutableTag');
    const tagScalarTokenAndValue = getStringAndScalarTokenFromMap(
      dockerImageBlock,
      'tag',
    );

    if (trackMutableTag && dockerImageRepository && tagScalarTokenAndValue) {
      trackables.push({
        trackMutableTag,
        dockerImageRepository,
        tag: tagScalarTokenAndValue.value,
        tagScalarTokenWriter: new ScalarTokenWriter(
          tagScalarTokenAndValue.scalarToken,
          doc.schema,
        ),
      });
    }
  }

  return trackables;
}

async function checkTagsAgainstArtifactRegistryAndModifyScalars(
  trackables: Trackable[],
  dockerRegistryClient: DockerRegistryClient,
): Promise<void> {
  for (const trackable of trackables) {
    const prefix = `${trackable.trackMutableTag}---`;

    const equivalentTags = (
      await dockerRegistryClient.getAllEquivalentTags({
        dockerImageRepository: trackable.dockerImageRepository,
        tag: trackable.trackMutableTag,
      })
    ).filter((t) => t.startsWith(prefix));

    // We assume that all the tags with the triple-dash in them are immutable:
    // once they point at a particular SHA, they never change. (Whereas the tag
    // we're "tracking" is mutable.)
    //
    // If the current tag has the right format for an immutable tag and it
    // points to the same image as the mutable tag, leave it alone: there's no
    // reason to create a no-op diff.
    if (equivalentTags.includes(trackable.tag)) {
      core.info(
        `for image ${trackable.dockerImageRepository}:${trackable.trackMutableTag}, preserving current tag ${trackable.tag}`,
      );
      continue;
    }
    // We can choose *any* of these equivalent triple-dashed tags, and it will
    // select the correct image version.
    //
    // Our tag structure increase over time (by including the number of commits
    // since the start as determined by `git rev-list --first-parent --count
    // HEAD`), and also includes the git commit at which it was built.
    //
    // So by choosing the lexicographically earliest of the equivalent tags, we
    // are most likely to choose a tag where the named git commit actually is
    // the commit that made a relevant change that affected the image.
    //
    // Additionally, this means that reverting a code change is likely to result
    // in a revert of the tag.  Imagine that tag `main---00123-abcd` is
    // currently running in both staging and prod, and a code change moves
    // `main` to `main---00130-bcde` and this is deployed to staging (opening a
    // prod promotion PR). A bug is found, so the code change X is reverted.
    // Reproducible builds will mean that the newest tag `main---00134-dcba`
    // will hopefully point to the same image version as `main---00123-abcd`.
    // Choosing the min version here will mean that we will in fact "revert"
    // staging to `main---00123-abcd`. This is now the exact same tag that is
    // running in prod, so the prod promotion PR can auto-close rather than
    // encouraging us to consider a no-op deploy to prod.
    const earliestMatchingTag = min(equivalentTags);
    if (!earliestMatchingTag) {
      throw new Error(
        `No tags on ${trackable.dockerImageRepository} start with '${prefix}'`,
      );
    }

    // It's OK if the current one is null because that's what we're overwriting, but we shouldn't
    // overwrite *to* something that doesn't exist.
    core.info(
      `for image ${trackable.dockerImageRepository}:${trackable.trackMutableTag}, changing to minimal matching tag ${earliestMatchingTag}`,
    );
    trackable.tagScalarTokenWriter.write(earliestMatchingTag);
  }
}
