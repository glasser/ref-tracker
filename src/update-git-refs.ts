import * as core from '@actions/core';
import * as yaml from 'yaml';
import { GitHubClient } from './github';
import {
  ScalarTokenWriter,
  getStringAndScalarTokenFromMap,
  getStringValue,
  getTopLevelBlocks,
} from './yaml';

interface Trackable {
  trackMutableRef: string;
  repoURL: string;
  path: string;
  ref: string;
  refScalarTokenWriter: ScalarTokenWriter;
}

export async function updateGitRefs(
  contents: string,
  gitHubClient: GitHubClient,
): Promise<string> {
  return core.group('Processing trackMutableRef', async () => {
    // The yaml module lets us parse YAML into three layers of abstraction:
    // - It can create raw JS arrays/objects/etc, which is simple to dealing
    //   with but loses track of everything relating to formatting.
    // - It can create a low-level "Concrete Syntax Tree" (CST) which lets us
    //   re-create the original document with byte-by-byte accuracy, but is
    //   awkward to read from (eg, you have to navigate maps item by item rather
    //   than using keys).
    // - It can create a high-level "Abstract Syntax Tree" (AST) which is easier
    //   to read from but loses some formatting details.
    //
    // We'd prefer to read ASTs and write CSTs, and in fact the module lets us
    // do exactly that. We first create CSTs with the "Parser". We then convert
    // it into ASTs with the Composer, passing in the `keepSourceTokens` option
    // which means that every node in the AST will have a `srcToken` reference
    // to the underlying CST node that created it. When we want to make changes,
    // we do that by writing to the CST node found in a `srcToken` reference.
    // Finally, when we're done, we stringify the CSTs (which have been mutated)
    // rather than the ASTs.
    core.info('Parsing');
    const topLevelTokens = [...new yaml.Parser().parse(contents)];
    const documents = [
      ...new yaml.Composer({ keepSourceTokens: true }).compose(topLevelTokens),
    ];

    // These files are all Helm values.yaml files, and Helm doesn't support a
    // multiple-document stream (with ---) for its value files. Or well, it
    // ignores any documents after the first, so there's no point in allowing
    // folks to put them in our codebase.
    if (documents.length > 1) {
      throw new Error('Multiple documents in YAML file');
    }

    // If the file is empty (or just whitespace or whatever), that's fine; we
    // can just leave it alone.
    if (documents.length < 1) {
      return contents;
    }

    const document = documents[0];

    core.info('Looking for trackMutableRef');
    const trackables = findTrackables(document);

    core.info('Checking refs against GitHub');
    await checkRefsAgainstGitHubAndModifyScalars(trackables, gitHubClient);
    core.info('Stringifying');
    return topLevelTokens
      .map((topLevelToken) => yaml.CST.stringify(topLevelToken))
      .join('');
  });
}

function findTrackables(doc: yaml.Document.Parsed): Trackable[] {
  const trackables: Trackable[] = [];

  // FIXME figure out error handling

  const { blocks, globalBlock } = getTopLevelBlocks(doc);

  let globalRepoURL: string | null = null;
  let globalPath: string | null = null;

  if (globalBlock?.has('gitConfig')) {
    const gitConfigBlock = globalBlock.get('gitConfig');
    if (!yaml.isMap(gitConfigBlock)) {
      throw Error('Document has `global.gitConfig` that is not a map');
    }
    // Read repoURL and path from 'global' (keeping them null if they're not
    // there, though throwing if they're there as non-strings).
    globalRepoURL = getStringValue(gitConfigBlock, 'repoURL');
    globalPath = getStringValue(gitConfigBlock, 'path');
  }

  for (const [key, value] of blocks) {
    if (!value.has('gitConfig')) {
      continue;
    }
    const gitConfigBlock = value.get('gitConfig');
    if (!yaml.isMap(gitConfigBlock)) {
      throw Error(`Document has \`${key}.gitConfig\` that is not a map`);
    }

    const repoURL = getStringValue(gitConfigBlock, 'repoURL') ?? globalRepoURL;
    const path = getStringValue(gitConfigBlock, 'path') ?? globalPath;
    const trackMutableRef = getStringValue(gitConfigBlock, 'trackMutableRef');
    const refScalarTokenAndValue = getStringAndScalarTokenFromMap(
      gitConfigBlock,
      'ref',
    );

    if (trackMutableRef && repoURL && path && refScalarTokenAndValue) {
      trackables.push({
        trackMutableRef,
        repoURL,
        path,
        ref: refScalarTokenAndValue.value,
        refScalarTokenWriter: new ScalarTokenWriter(
          refScalarTokenAndValue.scalarToken,
          doc.schema,
        ),
      });
    }
  }

  return trackables;
}

async function checkRefsAgainstGitHubAndModifyScalars(
  trackables: Trackable[],
  gitHubClient: GitHubClient,
): Promise<void> {
  for (const trackable of trackables) {
    // FIXME error handling
    const mutableRefCurrentSHA = await gitHubClient.resolveRefToSha({
      repoURL: trackable.repoURL,
      ref: trackable.trackMutableRef,
    });

    if (trackable.ref === trackable.trackMutableRef) {
      // The mutable ref was written down in ref too. We always want to replace
      // that with the SHA (and if we do the path-based check below we won't,
      // because they're the same). This is something that might happen when
      // you're first adding an app (ie just writing the same thing twice and
      // letting the automation "correct" it to a SHA).
      trackable.refScalarTokenWriter.write(mutableRefCurrentSHA);
      continue;
    }

    if (trackable.ref === mutableRefCurrentSHA) {
      // The thing we would write is already in the file.
      continue;
    }

    // OK, we've got a SHA that we could overwrite the current ref
    // (`trackable.ref`) with in the config file. But we don't want to do this
    // if it would be a no-op. Let's check the tree SHA
    // (https://git-scm.com/book/en/v2/Git-Internals-Git-Objects#_tree_objects)
    // at the given path to see if it has changed between `trackable.ref` and
    // the SHA we're thinking about replacing it with.
    const currentTreeSHA = await gitHubClient.getTreeSHAForPath({
      repoURL: trackable.repoURL,
      ref: trackable.ref,
      path: trackable.path,
    });
    const newTreeSHA = await gitHubClient.getTreeSHAForPath({
      repoURL: trackable.repoURL,
      ref: mutableRefCurrentSHA,
      path: trackable.path,
    });
    if (newTreeSHA === null) {
      throw Error(
        `Could not get tree SHA for ${mutableRefCurrentSHA} in ${trackable.repoURL} for ref ${trackable.path}`,
      );
    }
    // It's OK if the current one is null because that's what we're overwriting, but we shouldn't
    // overwrite *to* something that doesn't exist.
    core.info(
      `for path ${trackable.path}, got tree shas ${currentTreeSHA} for ${trackable.ref} and ${newTreeSHA} for ${mutableRefCurrentSHA}`,
    );
    if (currentTreeSHA === newTreeSHA) {
      core.info('(unchanged)');
    } else {
      core.info('(changed!)');
      trackable.refScalarTokenWriter.write(mutableRefCurrentSHA);
    }
  }
}
