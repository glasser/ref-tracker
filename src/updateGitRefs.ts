import * as core from '@actions/core';
import { CST, Parser } from 'yaml';
import { GitHubClient } from './github';

type CSTScalarToken = CST.FlowScalar | CST.BlockScalar;

interface Trackable {
  trackMutableRef: string;
  repoURL: string;
  path: string;
  ref: string;
  refScalarToken: CSTScalarToken;
}

// Returns null if the value isn't there at all; throws if it's there but isn't
// a string.
function getStringValue(
  node: CST.BlockMap | CST.FlowCollection,
  key: string,
): string | null {
  const scalarToken = getScalarTokenFromMap(node, key);
  if (scalarToken === null) {
    return null;
  }
  // FIXME this can throw
  return CST.resolveAsScalar(scalarToken).value;
}

function getScalarTokenFromMap(
  node: CST.BlockMap | CST.FlowCollection,
  key: string,
): CSTScalarToken | null {
  for (const item of node.items) {
    if (!CST.isScalar(item.key)) {
      continue;
    }
    // FIXME this can throw
    const keyScalar = CST.resolveAsScalar(item.key);
    if (keyScalar.value !== key) {
      continue;
    }
    if (!CST.isScalar(item.value)) {
      throw Error(`Value associated with ${key} is not a scalar`);
    }
    return item.value;
  }
  return null;
}

function findTrackables(doc: CST.Document): Trackable[] {
  const trackables: Trackable[] = [];

  // FIXME figure out error handling

  CST.visit(doc, ({ key, value }) => {
    if (
      !value ||
      (value.type !== 'block-map' && value.type !== 'flow-collection')
    ) {
      return;
    }

    const trackMutableRef = getStringValue(value, 'trackMutableRef');
    const repoURL = getStringValue(value, 'repoURL');
    const path = getStringValue(value, 'path');
    const refScalarToken = getScalarTokenFromMap(value, 'ref');
    const ref = refScalarToken
      ? CST.resolveAsScalar(refScalarToken).value
      : null;
    if (trackMutableRef && repoURL && path && refScalarToken && ref) {
      trackables.push({
        trackMutableRef,
        repoURL,
        path,
        ref,
        refScalarToken,
      });
    }
  });

  return trackables;
}

export async function updateGitRefs(
  contents: string,
  gitHubClient: GitHubClient,
): Promise<string> {
  return core.group('Processing trackMutableRef', async () => {
    core.info('Parsing');
    const topLevelTokens = [...new Parser().parse(contents)];

    core.info('Looking for trackMutableRef');
    const trackables = topLevelTokens.flatMap((topLevelToken) =>
      topLevelToken.type === 'document' ? findTrackables(topLevelToken) : [],
    );

    core.info('Checking refs against GitHub');
    await checkRefsAgainstGitHubAndModifyScalars(trackables, gitHubClient);
    core.info('Stringifying');
    return topLevelTokens.map((token) => CST.stringify(token)).join('');
  });
}

function setStringValue(scalarToken: CSTScalarToken, value: string) {
  const alreadyQuoted =
    scalarToken.type === 'single-quoted-scalar' ||
    scalarToken.type === 'double-quoted-scalar' ||
    scalarToken.type === 'block-scalar';
  CST.setScalarValue(scalarToken, value, {
    // We're working on the CST, not the AST, so there's no schema, which means
    // that the yaml library doesn't understand the difference between (say)
    // numbers and bools and strings. So it doesn't let you say "hey, try to
    // keep this plain (unquoted), unless it looks like a number". So we're
    // going to make sure everything we write ends up quoted. That said, if it's
    // already quoted, we'll keep the quote style rather than normalizing to
    // only one style of quotes (passing `type: undefined` does that).
    type: alreadyQuoted ? undefined : 'QUOTE_SINGLE',
  });
}

async function checkRefsAgainstGitHubAndModifyScalars(
  trackables: Trackable[],
  gitHubClient: GitHubClient,
): Promise<void> {
  for (const trackable of trackables) {
    // FIXME error handling
    const mutableRefCurrentSHA = await gitHubClient.resolveRefToSha({
      repoURL: trackable.repoURL,
      ref: trackable.trackMutableRef,
    });

    if (trackable.ref === trackable.trackMutableRef) {
      // The mutable ref was written down in ref too. We always want to replace
      // that with the SHA (and if we do the path-based check below we won't,
      // because they're the same). This is something that might happen when
      // you're first adding an app (ie just writing the same thing twice and
      // letting the automation "correct" it to a SHA).
      setStringValue(trackable.refScalarToken, mutableRefCurrentSHA);
      continue;
    }

    if (trackable.ref === mutableRefCurrentSHA) {
      // The thing we would write is already in the file.
      continue;
    }

    // OK, we've got a SHA that we could overwrite the current ref
    // (`trackable.ref`) with in the config file. But we don't want to do this
    // if it would be a no-op. Let's check the tree SHA
    // (https://git-scm.com/book/en/v2/Git-Internals-Git-Objects#_tree_objects)
    // at the given path to see if it has changed between `trackable.ref` and
    // the SHA we're thinking about replacing it with.
    const currentTreeSHA = await gitHubClient.getTreeSHAForPath({
      repoURL: trackable.repoURL,
      ref: trackable.ref,
      path: trackable.path,
    });
    const newTreeSHA = await gitHubClient.getTreeSHAForPath({
      repoURL: trackable.repoURL,
      ref: mutableRefCurrentSHA,
      path: trackable.path,
    });
    if (newTreeSHA === null) {
      throw Error(
        `Could not get tree SHA for ${mutableRefCurrentSHA} in ${trackable.repoURL} for ref ${trackable.path}`,
      );
    }
    // It's OK if the current one is null because that's what we're overwriting, but we shouldn't
    // overwrite *to* something that doesn't exist.
    core.info(
      `for path ${trackable.path}, got tree shas ${currentTreeSHA} for ${trackable.ref} and ${newTreeSHA} for ${mutableRefCurrentSHA}`,
    );
    if (currentTreeSHA === newTreeSHA) {
      core.info('(unchanged)');
    } else {
      core.info('(changed!)');
      setStringValue(trackable.refScalarToken, mutableRefCurrentSHA);
    }
  }
}
