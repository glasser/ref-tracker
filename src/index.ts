import * as core from '@actions/core';
import * as github from '@actions/github';
import * as glob from '@actions/glob';
import { throttling } from '@octokit/plugin-throttling';
import { readFile, writeFile } from 'fs/promises';
import { ArtifactRegistryDockerRegistryClient } from './artifactRegistry';
import { OctokitGitHubClient } from './github';
import { updateDockerTags } from './update-docker-tags';
import { updateGitRefs } from './update-git-refs';
import { updatePromotedValues } from './update-promoted-values';

/**
 * The main function for the action.
 * @returns {Promise<void>} Resolves when the action is complete.
 */
export async function main(): Promise<void> {
  try {
    const files = core.getInput('files');
    const globber = await glob.create(files);
    const filenames = await globber.glob();
    for (const filename of filenames) {
      await processFile(filename);
    }
  } catch (error) {
    // Fail the workflow run if an error occurs
    if (error instanceof Error) core.setFailed(error.message);
  }
}

async function processFile(filename: string): Promise<void> {
  return core.group(`Processing ${filename}`, async () => {
    let contents = await readFile(filename, 'utf-8');

    if (core.getBooleanInput('update-git-refs')) {
      const githubToken = core.getInput('github-token');
      const octokit = github.getOctokit(githubToken, throttling);
      const gitHubClient = new OctokitGitHubClient(octokit);
      contents = await updateGitRefs(contents, gitHubClient);
    }

    const artifactRegistryRepository = core.getInput(
      'update-docker-tags-for-artifact-registry-repository',
    );

    if (artifactRegistryRepository) {
      const dockerRegistryClient = new ArtifactRegistryDockerRegistryClient(
        artifactRegistryRepository,
      );
      contents = await updateDockerTags(contents, dockerRegistryClient);
    }

    if (core.getBooleanInput('update-promoted-values')) {
      contents = await updatePromotedValues(
        contents,
        core.getInput('promotion-target-regexp') || null,
      );
    }

    await writeFile(filename, contents);
  });
}

main();
