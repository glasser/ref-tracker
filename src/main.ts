import * as core from '@actions/core';
import * as github from '@actions/github';
import * as glob from '@actions/glob';
import { readFile, writeFile } from 'fs/promises';
import { GitHubClient, OctokitGitHubClient } from './github';
import { updateGitRefs } from './update-git-refs';
import { updatePromotedValues } from './update-promoted-values';

/**
 * The main function for the action.
 * @returns {Promise<void>} Resolves when the action is complete.
 */
export async function main(): Promise<void> {
  try {
    const githubToken = core.getInput('github-token');
    const octokit = github.getOctokit(githubToken);
    // FIXME consider adding @octokit/plugin-throttling
    const gitHubClient = new OctokitGitHubClient(octokit);
    const files = core.getInput('files');
    const globber = await glob.create(files);
    const filenames = await globber.glob();
    for (const filename of filenames) {
      core.info(`Looking at ${filename}`);
      await processFile(filename, gitHubClient);
    }
  } catch (error) {
    // Fail the workflow run if an error occurs
    if (error instanceof Error) core.setFailed(error.message);
  }
}

export async function processFile(
  filename: string,
  gitHubClient: GitHubClient,
): Promise<void> {
  return core.group(`Processing ${filename}`, async () => {
    core.info('Reading');
    let contents = await readFile(filename, 'utf-8');

    if (core.getBooleanInput('update-git-refs')) {
      contents = await updateGitRefs(contents, gitHubClient);
    }

    if (core.getBooleanInput('update-promoted-values')) {
      contents = await updatePromotedValues(
        contents,
        core.getInput('promotion-target-regexp') || null,
      );
    }

    await writeFile(filename, contents);
  });
}
