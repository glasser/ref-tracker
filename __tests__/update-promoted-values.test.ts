/**
 * Unit tests for the action's main functionality, src/main.ts
 *
 * These should be run as if the action was called from a workflow.
 * Specifically, the inputs listed in `action.yml` should be set as environment
 * variables following the pattern `INPUT_<INPUT_NAME>`.
 */

import { readFile } from 'fs/promises';
import { join } from 'path';
import { updatePromotedValues } from '../src/update-promoted-values';

// FIXME figure out logging (eg pass in a fake implementation of core
// to code, or mock it, or capture process.stdout, etc)
// Mock the GitHub Actions core library
// let infoMock: jest.SpyInstance;

describe('action', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    //infoMock = jest.spyOn(core, 'info').mockImplementation();
  });

  it('updates git refs', async () => {
    const contents = await readFile(
      join(__dirname, '__fixtures__', 'update-promoted-values', 'sample.yaml'),
      'utf-8',
    );
    const newContents = await updatePromotedValues(contents, 'prod');
    expect(newContents).toMatchSnapshot();

    // It should be idempotent in this case.
    expect(await updatePromotedValues(newContents, 'prod')).toBe(newContents);
  });
});
