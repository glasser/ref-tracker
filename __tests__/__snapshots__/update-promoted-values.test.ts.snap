// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action updates git refs 1`] = `
"some-service-staging:
  source:
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world
    trackMutableRef: main
    ref:        "abcdef1234567"

some-service-prod:
  promote:
    from: some-service-staging
    values: [[source, ref]]
  source:
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world
    ref:    abcdef1234567      # please

some-service-not-selected:
  promote:
    from: some-service-staging
    values: [[source, ref]]
  source:
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world
    ref:    overwrite-me      # this will not be updated because of the regexp
"
`;
